File: 2019/day02/python/solution.py
def format_input(input: str) -> list[int]:
    return [int(num) for num in input.strip().split(",") if num]


def part1(input: str, bypass_sub: bool = False) -> int:
    program = format_input(input)

    if program[1] != 9 and not bypass_sub:  # do not change the example provided
        program[1] = 12
        program[2] = 2

    for i in range(0, len(program), 4):
        opcode = program[i]
        if opcode == 99:
            break

        input1_index = program[i + 1]
        input2_index = program[i + 2]
        insertion_index = program[i + 3]

        if opcode == 1:
            program[insertion_index] = program[input1_index] + program[input2_index]
        elif opcode == 2:
            program[insertion_index] = program[input1_index] * program[input2_index]

    return program[0]


def part2(input):
    desired_outcome = 19690720

    for x in range(0, 100):
        for y in range(0, 100):
            program = format_input(input)
            program[1] = x
            program[2] = y

            if desired_outcome == part1(",".join(map(str, program)), True):
                return 100 * x + y

File: 2019/day02/javascript/solution.js
function formatInput(input) {
  return input.split(",").map(Number);
}

function part1(input) {
  const program = formatInput(input);

  for (let o = 0; o < program.length; o += 4) {
    const optcode = program[o];

    if (optcode === 99) break;

    const r1 = program[o + 1];
    const r2 = program[o + 2];
    const i = program[o + 3];

    if (optcode === 1) {
      program[i] = program[r1] + program[r2];
    } else if (optcode === 2) {
      program[i] = program[r1] * program[r2];
    }
  }
  return program[0];
}

function part2(input) {
  return undefined;
}

module.exports = { part1, part2 };
